src/_main_/utils/translation/__init__.py

```diff
@@ -106,46 +136,67 @@ def get_flattened_dict(self):
 
     def get_flattened_dict_for_excluded_keys(self):
         return self._excluded
-    
 
-    def translate(self, source_language: str, destination_language: str):
+
+    def translate (self, source_language: str, destination_language: str) -> Tuple[dict, List[str], List[dict]]:
         """
         Translate the flattened dictionary values from source_language to destination_language.
 
         Returns:
             dict: The translated dictionary in its original nested structure.
         """
-        # Flattened dictionary keys and values
-        keys = []
-        untranslated_text_entries = []
-        for key,value in self._flattened.items():
+
+        keys = [] # Flattened dictionary keys and values
+        untranslated_text_entries = [] # texts to be translated
+        text_hashes = [] # text hashes to be translated
+
+        for key, value in self._flattened.items():
+            if not value:
+                continue
+
+            # Check if the value should be excluded from translation
+            text_hash_exists, err = self.textHasService.text_hash_exists(value)
+            if err:
+                log.error("Error checking if text hash exists for", value, ":", err)
+                continue
+
+            if text_hash_exists:
+                continue
+
+            text_hash, err = self.textHasService.create_text_hash(args = { "text": value })
```

Comment:
don't we rather want to finish translating then we create the entries for the hash and the translations?  In fact.  can we fan out another thread to do this and not have it dependent on this main thread?  

Actually maybe a better idea is to add it to a queue again to do the processing.  so we don't have to wait for a response here.  the worst case is that we don't store the hash and so we have to translate it again.  so we should not block on the result from this.